### PR TITLE
hook/prot: Connectivity Map

### PR DESCRIPTION
**Pull request by jjhursey**
* This PR requires that PR #2773 be merged in - since this is a component of that framework.
   - This PR includes the `hook` framework PR.
   - Only the top-most commit should be reviewed for discussion
 * The goal of this PR is to expose this mechanism for discussion
 * See Issue #30 for more discussion

### CODE REVIEW

**Comment by artpol84** [comment]
bot:mellanox:retest

**Comment by jjhursey** [comment]
Some notes on the tunable options and output.

 * `-mca hook_prot_verbose VALUE`
     - General component verbosity
 * `-mca hook_prot_enable_mpi_init BOOL`
     - Enable map display at the bottom of `MPI_Init`
 * `-mca hook_prot_enable_mpi_finalize BOOL`
     - Enable map display at the top of `MPI_Finalize`
 * `-mca hook_prot_platform_prot VALUE`
      -  Alias environment variable: `MPI_PROT`
      - `1 : Same as -mca hook_prot_enable_mpi_init t`
      - `2 : Same as -mca hook_prot_enable_mpi_finalize t`
 * Environment variable: `MPI_PROT_MAX`
      - When to stop printing the prot table

Output lower than or equal to  `MPI_PROT_MAX`
```
     host | 0    1    2    3    4    5    6    7    8
    ======|==============================================
        0 : shm  ib   ib   ib   ib   ib   ib   ib   ib
        1 : ib   shm  ib   ib   ib   ib   ib   ib   ib
        2 : ib   ib   self ib   ib   ib   ib   ib   ib
        3 : ib   ib   ib   self ib   ib   ib   ib   ib
        4 : ib   ib   ib   ib   self ib   ib   ib   ib
        5 : ib   ib   ib   ib   ib   self ib   ib   ib
        6 : ib   ib   ib   ib   ib   ib   self ib   ib
        7 : ib   ib   ib   ib   ib   ib   ib   self ib
        8 : ib   ib   ib   ib   ib   ib   ib   ib   self
```

Output above `MPI_PROT_MAX` to `3*MMPI_PROT_MAX`:
```
     host | 0 1 2 3 4       8
    ======|====================
        0 : A C C C C C C C C
        1 : C A C C C C C C C
        2 : C C B C C C C C C
        3 : C C C B C C C C C
        4 : C C C C B C C C C
        5 : C C C C C B C C C
        6 : C C C C C C B C C
        7 : C C C C C C C B C
        8 : C C C C C C C C B
    key: A == shm
    key: B == self
    key: C == ib
```

File: ompi/mca/hook/prot/hook_prot_fns.c [hunk 1]
```diff
@@ -0,0 +1,803 @@
...
+
+    if (hpmp_myrank == 0) {
+        max2Dprotptr = getenv("MPI_PROT_MAX");
+        if (max2Dprotptr) {
+            max2Dprottable = atoi(max2Dprotptr);
+            max2D1Cprottable = 3 * max2Dprottable;
+        }
+
+        hostidprotptr = getenv("MPI_PROT_BRIEF");
+        if (hostidprotptr) { hostidprotbrief = atoi(hostidprotptr); }
```
Comment (jjhursey) [inline ompi/mca/hook/prot/hook_prot_fns.c]: @markalle It looks like `MPI_PROT_BRIEF` value is not used. Was the intention to force the "brief" output for all -prot output?

**Comment by jjhursey** [comment]
Ref PR https://github.com/open-mpi/ompi/pull/1974

**Comment by gpaulsen** [comment]
Also, there are some more details from face to face at the bottom of the [face to face minutes](https://github.com/open-mpi/ompi/wiki/Meeting-Minutes-2017-01)

**Comment by markalle** [comment]
For what it's worth the part they really want is    9a908f5b9bddd6675621c784c832441533cb4f73
that changes "unic" to "usnic"
Then the extra change    8f775b1f3a1ca856791164f5b971a2de274f4733was just for fun, because I thought it was neat to be able to fit larger tables on the screen.
 
Mark
 
----- Original message -----From: Geoff Paulsen <notifications@github.com>To: open-mpi/ompi <ompi@noreply.github.com>Cc: Mark Allen/Dallas/IBM@IBMUS, Mention <mention@noreply.github.com>Subject: Re: [open-mpi/ompi] hook/prot: Connectivity Map (#2825)Date: Thu, Jan 26, 2017 1:10 PM 
Also, there are some more details from face to face at the bottom of the face to face minutes

**Comment by jjhursey** [comment]
I think we might be able to use something like PR #1974 to get the string associated with the component. However, I think we might need some discussion about how the community feels about the section of this PR where we get the list of components actually being used between rank pairs.

Currently, we added a call to `pml/ob1` and `pml/cm` to get that information. Do folks think that is fine, or are there alternative suggestions?

**Comment by hppritcha** [comment]
bot:lanl:retest

**Comment by hppritcha** [comment]
the LANL dlopen-disable seems to have found a legit problem with this PR:

```
make[2]: Entering directory `/home/hppritcha/jenkins/workspace/ompi_master_pr_disable_dlopen/ompi/tools/ompi_info'
  CC       ompi_info.o
  CC       param.o
  GENERATE ompi_info.1
  CCLD     ompi_info
../../../ompi/.libs/libmpi.so: undefined reference to `dlsym'
collect2: error: ld returned 1 exit status
make[2]: *** [ompi_info] Error 1
make[2]: Leaving directory `/home/hppritcha/jenkins/workspace/ompi_master_pr_disable_dlopen/ompi/tools/ompi_info'
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory `/home/hppritcha/jenkins/workspace/ompi_master_pr_disable_dlopen/ompi'
make: *** [install-recursive] Error 1
```

**Comment by jjhursey** [comment]
@hppritcha Yeah, you are right. We need to do some work on this PR anyway before it's ready to review. We'll note that error for fixing during the next revision.

**Comment by hppritcha** [comment]
bot:lanl:retest

**Comment by jjhursey** [comment]
This PR needs some significant work before it's ready to merge. So I wouldn't worry too much about the CI until that work has started.

**Comment by hppritcha** [comment]
bot:lanl:retest

**Comment by jjhursey** [comment]
Now that the hook framework has been committed to `master` I need to update this branch. I'll do that today.

**Comment by jjhursey** [comment]
The branch has been rebased onto master. The two commits represent the `prot` component.

There is still work to do on this feature - so I'm keeping the WIP label on it.

**Comment by gpaulsen** [comment]
:retest:

**Comment by bwbarrett** [comment]
@jjhursey are you looking to add this to the 3.0 release?

**Comment by jjhursey** [comment]
It still needs some work. I don't think it'll make it for v3.0, but _should_ be ready for the release that follows.

**Comment by gpaulsen** [comment]
The community decided on a call 2 weeks ago, to NOT to take this for v3.0, but to aim for v3.1.

**Comment by bwbarrett** [comment]
Thanks, @gpaulsen.  Apparently, I have no short term memory, hence writing everything down...

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/73ca9ce0ceed7b37f7d60da3250d7259

**Comment by jjhursey** [comment]
I'm going to close this PR for now. @gpaulsen @markalle Let's re-open a new one once we have finished refining. We should restart that conversation as I forget what is left to do.

**Comment by gpaulsen** [comment]
This Functionality was later re-implemented for the v5.0 timeframe in https://github.com/open-mpi/ompi/pull/5507

**Pull request closed without merging**